src/Transaction.cc

```diff
@@ -0,0 +1,293 @@
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "ClientTransactionTask.h"
+#include "Transaction.h"
+
+namespace RAMCloud {
+
+/**
+ * Constructor for a transaction.
+ *
+ * \param ramcloud
+ *      Overall information about the calling client.
+ */
+Transaction::Transaction(RamCloud* ramcloud)
+    : ramcloud(ramcloud)
+    , taskPtr(new ClientTransactionTask(ramcloud))
+    , commitStarted(false)
+{
+}
+
+/**
+ * Commits the transaction defined by the operations performed on this
+ * transaction (read, remove, write).  This method blocks until a decision is
+ * reached but not until the decisions are synced.
+ *
+ * \return
+ *      True if the transaction was able to commit.  False otherwise.
+ */
+bool
+Transaction::commit()
+{
+    if (!commitStarted) {
+        commitStarted = true;
+        ramcloud->transactionManager.addTransactionTask(taskPtr);
+    }
+
+    ClientTransactionTask* task = taskPtr.get();
+
+    while (task->getDecision() == WireFormat::TxDecision::INVALID) {
+        ramcloud->poll();
+    }
+
+    if (task->getStatus() != STATUS_OK) {
+        ClientException::throwException(HERE, task->getStatus());
+    }
+
+    return (task->getDecision() == WireFormat::TxDecision::COMMIT);
+}
+
+/**
+ * Block until the decision of this transaction commit is accepted by all
+ * participant servers.  If the commit has not yet occurred and a decision is
+ * not yet reached, this method will also start the commit.
+ */
+void
+Transaction::sync()
+{
+    if (!commitStarted) {
+        commitStarted = true;
+        ramcloud->transactionManager.addTransactionTask(taskPtr);
+    }
+
+    ClientTransactionTask* task = taskPtr.get();
+
+    while (!task->isReady()) {
+        ramcloud->poll();
+    }
+
+    if (task->getStatus() != STATUS_OK) {
+        ClientException::throwException(HERE, task->getStatus());
+    }
+}
+
+/**
+ * Commits the transaction defined by the operations performed on this
+ * transaction (read, remove, write).  This method blocks until a participant
+ * servers have accepted the decision.
+ *
+ * \return
+ *      True if the transaction was able to commit.  False otherwise.
+ */
+bool
+Transaction::commitAndSync()
+{
+    sync();
+    ClientTransactionTask* task = taskPtr.get();
+    return (task->getDecision() == WireFormat::TxDecision::COMMIT);
+}
+
+/**
+ * Read the current contents of an object as part of this transaction.
+ *
+ * \param tableId
+ *      The table containing the desired object (return value from
+ *      a previous call to getTableId).
+ * \param key
+ *      Variable length key that uniquely identifies the object within tableId.
```

Comment:
Let's start saying "primary key" instead of just "key", since we now have secondary indexes.